services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/signature/DccFeignDelegator.java

```diff
@@ -0,0 +1,58 @@
+package app.coronawarn.server.services.distribution.dgc.client.signature;
+
+import feign.Client;
+import feign.Request;
+import feign.Request.Options;
+import feign.Response;
+import feign.httpclient.ApacheHttpClient;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+
+public class DccFeignDelegator implements Client {
+
+  public static final String X_SIGNATURE = "X-SIGNATURE";
+
+  private final ApacheHttpClient apacheHttpClient;
+  private final DccSignatureValidator dccSignatureValidator;
+
+  public DccFeignDelegator(ApacheHttpClient apacheHttpClient, DccSignatureValidator dccSignatureValidator) {
+    this.apacheHttpClient = apacheHttpClient;
+    this.dccSignatureValidator = dccSignatureValidator;
+  }
+
+  @Override
+  public Response execute(Request request, Options options) throws IOException {
+    Response response = apacheHttpClient.execute(request, options);
+    String body = new String(response.body().asInputStream().readAllBytes());
+    String signature = getSignature(response, request.url());
+
+    dccSignatureValidator.checkSignature(signature, body);
+
+    // response recreated to avoid closed stream exception.
+    return Response.builder()
+        .status(response.status())
+        .reason(response.reason())
+        .headers(response.headers())
+        .request(response.request())
+        .body(body, StandardCharsets.UTF_8).build();
+  }
+
+  /**
+   * Get the signature from the HTTP response headers.
+   *
+   * @param response - response.
+   * @param requestUrl - request url.
+   * @return - signature found on 'x-signature' headers key.
+   * @throws IOException - thrown if 'x-signature' header is missing.
+   */
+  private String getSignature(Response response, String requestUrl) throws IOException {
+    Collection<String> header = response.headers().get(X_SIGNATURE);
```

Comment:
I think at some point Feign Headers were lowercased. Maybe we can double check this in a test

Reply:
@alexmoroianu please deploy the build resulting image on INT and test if the headers are coming in lowercase or uppercase...

Maybe it would be safer to search case insensitive for x-signature

Reply:
@hilmarf. I will deploy but I've looked into the code and the feign instantiate the response headers(map) like this: Util.caseInsensitiveCopyOf(builder.headers); . Although I debugged and this map returns the value no matter what case type I pass as an argument. It returns the header with an uppercase, lowercase or combined key.